### PR TITLE
#10539 extensions.rst

### PR DESCRIPTION
**Pull request by Konntroll**
Finished editing extensions.rst for issue #10539. Let me know if you'd like me to make any changes.

The example on line 189 doesn't produce any output and causes documentation build to fail unless you remove the code block on lines 82-84 of the "extensions_putting_together_ts.py" example file:

    this.override<Custom.Props>({
      margin: 5,
    })

Removing it seems to fix both issues. Not sure what this bit is supposed to do nor whether removing it breaks anything else.

### CODE REVIEW

**Comment by bryevdv** [comment]
@Konntroll what is the error that you are seeing when that is not deleted?

File: sphinx/source/docs/user_guide/examples/extensions_putting_together_ts.py [hunk 1]
```diff
@@ -66,22 +66,20 @@
   }
 
   static init_Custom(): void {
-    // If there is an associated view, this is typically boilerplate.
+    // If there is an associated view, it's typically a boilerplate.
```
Comment (bryevdv) [inline sphinx/source/docs/user_guide/examples/extensions_putting_together_ts.py:69]: I don't like this new wording, it's the code for the view that is typically boilerplate, and I don't think this conveys that exactly

**Comment by Konntroll** [comment]
Reverted to the original wording for view boilerplate, both in the code snippet within the chapter and in the example file.

As for the for the error message that I've been getting when building the docs, this goes as follows:
> Compilation failed:[ 91%] docs/user_guide/extensions
> source/docs/user_guide/examples/extensions_putting_together_ts.py:Custom.ts:76:19 - error TS2558: Expected 0 type arguments, but got 1.
> 76     this.override<Custom.Props>({
>                     ~~~~~~~~~~~~

**Comment by bryevdv** [comment]
@mattpap any idea about the error message? The usage looks fairly similar to other places in the codebase

**Comment by mattpap** [comment]
This is because bokehjs pre 2.3-dev1 is being used, most likely 2.2.

**Comment by bryevdv** [comment]
@mattpap that seems like a problem with CI then, but also this seems to imply there is a significant breaking change for custom extensions? I don't see anything at https://github.com/bokeh/bokeh/blob/branch-2.3/sphinx/source/docs/releases/2.3.0.rst

**Comment by bryevdv** [comment]
@Konntroll can you put the line back, and let it fail so we can examine the logs in CI?

**Comment by Konntroll** [comment]
Done. Let me know if you need more changes, @bryevdv. Now that it's weekend, I should have (more) time to work on this and other edits.

File: sphinx/source/docs/user_guide/examples/extensions_putting_together_ts.py [hunk 2]
```diff
@@ -69,19 +69,21 @@
...
       text:   [ String ],
       slider: [ Ref(Slider) ],
     }))
 
-    this.override<Custom.Props>({
-      margin: 5,
+    this.override<Custom.Props>({	
+      margin: 5,	
     })
+
```
Comment (mattpap) [inline sphinx/source/docs/user_guide/examples/extensions_putting_together_ts.py:86]: Whitespace & empty line.

File: sphinx/source/docs/user_guide/examples/extensions_putting_together_ts.py [hunk 3]
```diff
@@ -69,19 +69,21 @@
...
-    // should normally line up 1-1 with the Python model class. Most property
-    // types have counterparts, e.g. bokeh.core.properties.String will be
-    // ``p.String`` in the JS implementation. Any time the JS type system is not
-    // yet as complete, you can use ``p.Any`` as a "wildcard" property type.
+    // The this.define() block adds corresponding "properties" to the JS
+    // model. These should normally line up 1-1 with the Python model
+    // class. Most property types have counterparts. For example,
+    // bokeh.core.properties.String will correspond to ``String`` in the
+    // JS implementation. Where JS lacks a given type, you can use
+    // ``p.Any`` as a "wildcard" property type.
```
Comment (mattpap) [inline sphinx/source/docs/user_guide/examples/extensions_putting_together_ts.py:77]: This is explanation is out of date. `p.Any` (and similar) are deprecated, and one should use `Any` kind instead, e.g.:
```ts
this.define<Custom.Props>(({Any, String, Ref}) => ({
  text:   [ String ],
  slider: [ Ref(Slider) ],
  whatever: [ Any ],
}))
```
This also means that there is no direct correspondence between bokeh's `String` and bokehjs `String` (kind). The closest is bokeh's `String` and bokehjs' `p.PrimitiveProperty<String>(..., kinds.String, ...)` (I'm using namespace `kinds` to signify the difference between `String` kind and `String` JS built-in type.

**Comment by mattpap** [comment]
> (...) but also this seems to imply there is a significant breaking change for custom extensions?  I don't see anything at https://github.com/bokeh/bokeh/blob/branch-2.3/sphinx/source/docs/releases/2.3.0.rst

It is a breaking change, but most changes to core bokehjs are. Also any upgrade of TypeScript (minor versions in particular) is usually a breaking change. This definitively needs a migration note, but I was putting this off until all changes to the property system were done.

File: sphinx/source/docs/user_guide/examples/extensions_putting_together_ts.py (hunk 3 above)
Comment (Konntroll) [reply sphinx/source/docs/user_guide/examples/extensions_putting_together_ts.py:77]: Would the following be more appropriate, then?

    // ... Most property types have counterparts. For example,
    // bokeh.core.properties.String corresponds to
    // p.PrimitiveProperty<String>(..., kinds.String, ...) in the JS
    // implementation, where ``kinds`` is a custom namespace.
    // If JS lacks a given type, you can use ``Any`` as a "wildcard"
    // property type.

Let me know if this requires a more thorough rewrite, @mattpap.

**Review by bryevdv — APPROVED** [review]

**Pull request merged**